[new module] Added Azure SQL Servers support - cloud/azure/azure_rm_sql_servers

zikalino opened the pull request:
##### SUMMARY
Adding support for Azure SQL Servers

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
azure_rm_sql_servers

##### ANSIBLE VERSION
```
ansible 2.5.0 (devel 125c4b135d) last updated 2017/10/17 07:30:02 (GMT +000)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /ansible/lib/ansible
  executable location = /ansible/bin/ansible
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```

##### ADDITIONAL INFORMATION

Commit:
Added new module: Azure RM SQL Servers

Commit:
fixed sanity issues

Commit:
enabled integration test

Commit:
disabling test as Microsoft.Sql namespace not available

Commit:
renamed servers -> server

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sql_server.py:
```diff
@@ -0,0 +1,292 @@
+---
+module: azure_rm_sql_server
+version_added: "2.5"
+short_description: Manage an Server.
```
short_description should not have a `.`, are you able to expand on this like, `Manage a Azure SQL Server`?

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sql_server.py:
```diff
@@ -0,0 +1,292 @@
+version_added: "2.5"
+short_description: Manage an Server.
+description:
+    - Create, update and delete an instance of Server.
```
SQL

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sql_server.py:
```diff
@@ -0,0 +1,292 @@
+    location:
+        description:
+            - Resource location.
+        required: True
```
It would be nice if this wasn't required and could be derived from the resource group specified.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sql_server.py:
```diff
@@ -0,0 +1,292 @@
+    tags:
+        description:
+            - Resource tags.
+        required: False
```
`required: False` is the default and is usually omitted.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sql_server.py:
```diff
@@ -0,0 +1,292 @@
+        suboptions:
+            type:
+                description:
+                    - "The identity type. Set this to 'SystemAssigned' in order to automatically create and assign an Azure Active Directory principal for th
```
I would argue that `SystemAssigned` should be the default value if this is not set or can a SQL Server not have an identity? Is there a reason why this is a nested option under the `type` key? If not can identity just by the type value or do you see there being more options available to identity?

When quoting a value, you can enclose it with `C()`, e.g. `'SystemAssigned'` should be `C(SystemAssigned)`. You can also use `I()` when referring to option option names.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sql_server.py:
```diff
@@ -0,0 +1,292 @@
+                description:
+                    - "The identity type. Set this to 'SystemAssigned' in order to automatically create and assign an Azure Active Directory principal for th
+                       e resource. Possible values include: 'SystemAssigned'"
+                required: False
```
not needed

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sql_server.py:
```diff
@@ -0,0 +1,292 @@
+                    - "The identity type. Set this to 'SystemAssigned' in order to automatically create and assign an Azure Active Directory principal for th
+                       e resource. Possible values include: 'SystemAssigned'"
+                required: False
+    administrator_login:
```
Change this option to `admin_username` to align with the other Azure modules.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sql_server.py:
```diff
@@ -0,0 +1,292 @@
+        description:
+            - Administrator username for the server. Once created it cannot be changed.
+        required: False
+    administrator_login_password:
```
Change this to `administrator_password` to align with other Azure modules.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sql_server.py:
```diff
@@ -0,0 +1,292 @@
+    administrator_login:
+        description:
+            - Administrator username for the server. Once created it cannot be changed.
+        required: False
```
Not needed

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sql_server.py:
```diff
@@ -0,0 +1,292 @@
+    administrator_login_password:
+        description:
+            - The administrator login password (required for server creation).
+        required: False
```
Not needed

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sql_server.py:
```diff
@@ -0,0 +1,292 @@
+    version:
+        description:
+            - The version of the server.
+        required: False
```
Not needed

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sql_server.py:
```diff
@@ -0,0 +1,292 @@
+        required: False
+    version:
+        description:
+            - The version of the server.
```
Can you expand on this more, e.g. how the default is derived if omitted, where to find a list of valid versions and so on.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sql_server.py:
```diff
@@ -0,0 +1,292 @@
+        required: False
+    administrator_login_password:
+        description:
+            - The administrator login password (required for server creation).
```
Is this the same with `administrator_username`, can it not be changed? Should there be another parameter like `update_password` which only makes a change on creation or all the time, see the `user` module for more details http://docs.ansible.com/ansible/latest/user_module.html.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sql_server.py:
```diff
@@ -0,0 +1,292 @@
+'''
+
+EXAMPLES = '''
+      - name: Create (or update) Servers
```
Can you keep this left aligned (remove the 4 spaces), also try and avoid using `{{}}` blocks and put valid values in there, it makes it easier for newer people to see what the values can actually be.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sql_server.py:
```diff
@@ -0,0 +1,292 @@
+            type: "{{ type }}"
+          administrator_login: mylogin
+          administrator_login_password: Testpasswordxyz12!
+          version: "{{ version }}"
```
Would be nice if you can actually show an example of a valid version value, nothing worse than seeing that it can be specified but not knowing the form it takes.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sql_server.py:
```diff
@@ -0,0 +1,292 @@
+'''
+
+RETURN = '''
+state:
```
We should start looking at stopping this behaviour and returning relevant values in the root level of the return. When testing this module I see the following is returned

```
"state": {
    "administrator_login": "username",
    "fully_qualified_domain_name": "jborean-sql-server.database.windows.net",
    "id": "/subscriptions/3f7e29ba-24e0-42f6-8d9c-5149a14bda37/resourceGroups/jborean-testgroup/providers/Microsoft.Sql/servers/jborean-sql-server",
    "kind": "v12.0",
    "location": "australiaeast",
    "name": "jborean-sql-server",
    "state": "Ready",
    "type": "Microsoft.Sql/servers",
    "version": "12.0"
}
```

So let's do the following and create a return value for:
* fqdn
* id
* version

I don't think the other values are really necessary and are more suited towards a facts module if we want to gather that information. I decided to omit the `location`, `admin login`, `kind, `name`, `state` and `type` as they would be ones you set in the module parameters and already know. Version is a tiny bit different as you can omit it and I could argue `type` could potentially be returned if you can specify different types in the future.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sql_server.py:
```diff
@@ -0,0 +1,292 @@
+            ),
+            administrator_login_password=dict(
+                type='str',
+                required=False
```
Add another entry `no_log=True` to stop Ansible from logging this value, this should be set on any secret parameters (like passwords) to keep it a secret.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sql_server.py:
```diff
@@ -0,0 +1,292 @@
+    def exec_module(self, **kwargs):
+        """Main module execution method"""
+
+        for key in list(self.module_arg_spec.keys()) + ['tags']:
```
Have a look at [azure_rm_virtualmachine](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/azure/azure_rm_virtualmachine.py) to see how to do this, you usually set the parameter as None in the __init__ and these are populated in `AzureRMModuleBase` for you.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sql_server.py:
```diff
@@ -0,0 +1,292 @@
+        else:
+            self.log("Server instance already exists")
+            if self.state == 'absent':
+                self.delete_servers()
```
This shouldn't be run in check mode.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sql_server.py:
```diff
@@ -0,0 +1,292 @@
+                self.results['changed'] = True
+                self.log("Server instance deleted")
+            elif self.state == 'present':
+                self.log("Need to check if Server instance has to be deleted or may be updated")
```
You need some code to check if the server actually needs updating, right now the module is reporting a change when running it twice which is not idempotent.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sql_server.py:
```diff
@@ -0,0 +1,292 @@
+                self.results['state'] = self.create_update_servers()
+                self.results['changed'] = True
+            else:
+                self.results['state'] = response
```
We should look at changing the return values, see my note in the documentation portion.

jborean93 commented on test/integration/targets/azure_rm_sql_server/tasks/main.yml:
```diff
@@ -0,0 +1,13 @@
+- name: Create Servers
```
I know you were saying you were having issues trying to get this to run but we should also look at expanding these tests a lot more, usually we look at performing 3 tasks per action, e.g. 3 tasks for creating a server, 3 tasks for modifying a server and 3 tasks for removing a server. The 3 tasks per action should follow;

* Run the task in check mode and check that the server wasn't actually created
* Run the task normally and check that the server was created as expected
* Run the task again (idempotent) and check that a change wasn't reported

jborean93 reviewed (changes requested):
Have added some comments, some things we would need to do before merging is

* Fix up the idempotency, right now it reports a change everytime it runs
* Expand the tests to validate the module
* Make the return values more Ansible like

Overall the module works for me and creates a SQL Server which is great. One question I do have more on the Azure side, is it planned to add support for more SQL types, e.g. MySQL and so or is that not supported in Azure or going to be in separate modules?

Commit:
updates as requested in the review

Commit:
udpates to sql server

Commit:
renamed module

Commit:
replaced has_key by in

zikalino commented:
ready_for_review

zikalino commented:
fixed several major issues:
- renaming, etc.
- idempotence support
- added update / idempotence tests
I may have missed some smaller issues.
Please recheck if still any important changes needed

Commit:
flattened results

Commit:
final updates to sql server

zikalino commented:
ready_for_review

zikalino commented:
@jborean93 please check the module again, i think it's ready for merging.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sqlserver.py:
```diff
@@ -0,0 +1,321 @@
+        description:
+            - The name of the server.
+        required: True
+    tags:
```
I'm pretty sure you don't need the tags as you are extending the `azure_tags` documentation below.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sqlserver.py:
```diff
@@ -0,0 +1,321 @@
+    def exec_module(self, **kwargs):
+        """Main module execution method"""
+
+        for key in list(self.module_arg_spec.keys()) + ['tags']:
```
What's the point of this whole block, `AzureRMModuleBase` should be setting `self.x` based on the module_arg_spec

zikalino replied:
well, i have tried and it doesn't work as expected, so i have to leave it for now...

jborean93 replied:
Ok, will try and debug and find out what is happening, it should just work.

jborean93 replied:
Ok I see my mistake, most modules just have

```
for key in list(self.module_arg_spec.keys()) + ['tags']:
    setattr(self, key, kwargs[key])
```

Is there a reason why you have all these if statements as the above should be enough for you.

zikalino replied:
yes, these attributes are sticked into objects padded to Azure SDK, so treated in different way

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sqlserver.py:
```diff
@@ -0,0 +1,321 @@
+        self.mgmt_client = self.get_mgmt_svc_client(SqlManagementClient,
+                                                    base_url=self._cloud_environment.endpoints.resource_manager)
+
+        try:
```
No need for a `try/except` here as azure_rm_common handles a missing source group

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sqlserver.py:
```diff
@@ -0,0 +1,321 @@
+        except CloudError:
+            self.fail('resource group {0} not found'.format(self.resource_group))
+
+        if not ("location" in self.parameters):
```
This doesn't seem to be working, I got `The error was: msrest.exceptions.ValidationError: Parameter 'Server.location' can not be None` in an error which means this musn't work. Try doing;

```
if self.location is None:
    self.location = resource_group.location
```

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sqlserver.py:
```diff
@@ -0,0 +1,321 @@
+        self.parameters = dict()
+        self.parameters['identity'] = dict()
+
+        self.results = dict(changed=False, state=dict())
```
`state` doesn't need to be there.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sqlserver.py:
```diff
@@ -0,0 +1,321 @@
+                self.results['changed'] = True
+            else:
+                self.results['changed'] = old_response.__ne__(response)
+            self.results.update(response)
```
Instead of updating the results object here with the response and then removing the unneeded ones below, just set the results you want manually.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sqlserver.py:
```diff
@@ -0,0 +1,321 @@
+            self.log("Creation / Update done")
+        elif self.to_do == Actions.Delete:
+            self.log("SQL Server instance deleted")
+            self.delete_sqlserver()
```
Need to not run when in check mode

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_sqlserver.py:
```diff
@@ -0,0 +1,321 @@
+            self.results['changed'] = True
+        else:
+            self.log("SQL Server instance unchanged")
+            self.results['state'] = old_response
```
This shouldn't be returned, it should be the same values that you set in add or update.

jborean93 commented on test/integration/targets/azure_rm_sqlserver/aliases:
```diff
@@ -0,0 +1,2 @@
+cloud/azure
+destructive
```
You will need to add `posix/ci/cloud/group2/azure` to this list so the tests actually run in CI

jborean93 reviewed (changes requested):
I couldn't really test this as I came across the error

`fatal: [127.0.0.1]: FAILED! => {"changed": false, "msg": "Error creating the SQL Server instance: Azure Error: InvalidRequestContent\nMessage: The request content was invalid and could not be deserialized: 'Required property 'type' not found in JSON. Path 'identity', line 1, position 44.'."}`

A few things to note;

* The tests need to be activated by adding the alias I mentioned in the comment (that should have picked up the error I recieved)
* The tests need to be expanded to cover a check mode, normal and idempotent tests of the actions `create`, `update`, and `delete` scenarios
* I'm not sure if the tags are actually supported, I was unable to test it but couldn't find anything that really seem to check and set them
* The location is still required, it fails if you don't set it

Commit:
updates as requested in the review

zikalino commented:
@jborean93 I have made all necessary updates, seems like test can't run as Microsoft.Sql namespace has to be added.

Error creating the SQL Server instance: Azure Error: MissingSubscriptionRegistration\nMessage: The subscription is not registered to use namespace 'Microsoft.Sql'. See https://aka.ms/rps-not-found for how to register subscriptions.\nException Details:\n\tError Code: MissingSubscriptionRegistration\n\tMessage: The subscription is not registered to use namespace 'Microsoft.Sql'. See https://aka.ms/rps-not-found for how to register subscriptions.\n\tTarget: Microsoft.Sql

jborean93 commented:
Looks like the tests are failing because we aren't authorised in CI, will see if @mattclay can help.

jborean93 commented:
@zikalino once we get the tests running we need to look at expanding the tests, there is at least 1 issue I can find that the tests should have picked up if we had some more tests. Let me know if you are confused as to what I am asking for, I'm happy to discuss it further over IRC.

zikalino commented:
@jborean do you mean taking location from resource group? this could be difficult. some resources can't be created in particular location (I believe i couldn't create SQL server in the same location as the one where default rg is created).

jborean93 commented:
@zikalino, for the location part I see the benefit in being able to set the location independently to override the resource group but most modules seem to have it as an optional value and if not set explicitly it is derived from the resource group location.

What I am talking about is expanding the tests from just the create, update, delete tasks to test each scenario under check mode, normal and an idempotent task. In the end it would look similar to

```
- name: create sql server (check)
  azure_rm_sqlserver:
    ...
  register: create_server_check

- name: check if server exists (check)
  ... ? use a facts module if possible
  register: create_server_result_check

- name: assert create sql server results (check)
  assert:
    that:
    - create_server_check is changed
    - create_server_check.exists == False

- name: create sql server
  azure_rm_sqlserver:
    ...
  register: create_server

- name: check if server exists
  ... ? use a facts module if possible
  register: create_server_result

- name: assert create sql server results
  assert:
    that:
    - create_server is changed
    - create_server.exists == True

- name: create sql server (idempotent)
  azure_rm_sqlserver:
    ...
  register: create_server_again

- name: assert create sql server results (idempotent)
  assert:
    that:
    - not create_server_again is changed
```

This would be repeated again for updating a SQL server and then finally again for removing a SQL server.

jborean93 commented:
@zikalino, the Microsoft.SQL resource has been registered and the tests are now running.

zikalino commented:
@jborean93 seems now tests are passing. should i still add more tests to this pr? or is it ok to merge, and i could expand tests when adding facts module?

jborean93 commented:
I think we should still continue to add the extra tests just without that second step of checking the results with a fact module for now. We can still do some assertions on the results of just the task itself.

Do you know how to run tests locally?

Commit:
added more tests

Commit:
deleting unexisting instance up front

zikalino commented:
@jborean93 I have more tests for check_mode and also delete unexisting sql server test (which i moved to front, as delete after delete seems a bit unstable -- second delete may get invalid information that resource still exists)

jborean93 commented:
Looks better, we still need a test for each action again to ensure that no changes are recorded (idempotent).

Commit:
more tests

zikalino commented:
@jborean93 how about now? added 2 more tests

jborean93 commented on test/integration/targets/azure_rm_sqlserver/tasks/main.yml:
```diff
@@ -0,0 +1,105 @@
+      - output.changed
+      - output.state == 'Ready'
+
+- name: Create again instance of SQL Server -- check mode
```
Delete this one, you only need to run the check mode test for a scenario once

jborean93 commented on test/integration/targets/azure_rm_sqlserver/tasks/main.yml:
```diff
@@ -0,0 +1,105 @@
+    random_postfix: "{{ 1000 | random }}"
+  run_once: yes
+
+- name: Delete unexisting instance of SQL Server -- check mode
```
Take away this test

jborean93 commented on test/integration/targets/azure_rm_sqlserver/tasks/main.yml:
```diff
@@ -0,0 +1,105 @@
+    that:
+      - output.changed == false
+
+- name: Delete unexisting instance of SQL Server
```
Don't need this test as you should be starting with a random resource group so no SQL server would exist anyway

jborean93 reviewed (commented):
Looking good, left some comments on some tests, only thing remaining is to delete the instance again right at the end for idempotency checks

Commit:
another round of test changes

jborean93 committed:
removed uneeded test

zikalino commented:
shipit

jborean93 commented:
Awesome @zikalino, are you able to raise a PR to add this module to the changelog for 2.5 https://github.com/ansible/ansible/blob/devel/CHANGELOG.md. We try to keep things in alphabetical order in each section.
